3주차 과제 - 실습과제 3번 채널별 월 매출액 테이블 생성

mingyu-lee opened the pull request:
4번 코호트 과제까지 진행해보려고 합니다만, 일단 기본 과제 3가지 완료하여 PR 올립니다.

mingyu-lee committed:
3주차 과제 - 실습과제 3번 채널별 월 매출액 테이블 생성

mingyu-lee committed:
3주차 과제 결과 화면 html, md 추가

mingyu-lee commented:
@keeyong 
colab 샘플에서 sqlalchemy 사용하시는 것 보고 영감(?)을 받아 jupyter로 진행해봤습니다.
file 로 리뷰하긴 어려우실 듯 하여 결과 화면을 html과 markdown 파일로 만들어 같이 올렸습니다.

내일과 모레 시간이 된다면 4번도 한번 해보고 4번 PR 또 올려보겠습니다.

감사합니다.

keeyong commented:
1번 잘 하셨고 2, 3번 잘 하셨는데 RAW_DATA.SESSION_TRANSACTION 테이블의 필드 의미가 잘 전달이 안 된 것으로 보입니다. 제가 숙제를 내며 했던 정의는 REFUND  필드의 값이 False 이면 netRevenue가 0이 되고 grossRevenue는 amount가 되는 것입니다. 즉 순매출액의 경우에는 환불이 된 경우 0이 되는 것이죠. 그 부분을 제외하면 잘 하셨습니다.

keeyong commented:
제가 말씀드린 정의대로 2번 문제를 하면 아래와 같이 됩니다:

```
select
    USERID
  , TOTAL_REVENUE
from (
     select
         USERID
       , SUM(case when REFUNDED is false then AMOUNT end) as TOTAL_REVENUE
     from RAW_DATA.SESSION_TRANSACTION       ST
          join RAW_DATA.USER_SESSION_CHANNEL USC on USC.SESSIONID = ST.SESSIONID
     group by 1
     ) TOP_USER_REVENUE
order by 2 desc
limit 10;
```

keeyong commented:
3번 문제의 경우 paidUsers 계산이 아래와 같이 되어 있습니다:

> sum(case when AMOUNT is not null and REFUNDED is false then 1 else 0 end) as PAID_USERS

이를 아래와 같이 바꾸면 구매한 유니크한 사용자 수를 카운트할 수 있습니다.

> count(distinct case when AMOUNT is not null and REFUNDED is false then userid end) as PAID_USERS

mingyu-lee committed:
PAID_USERS, CONVERSION_RATE 개선
- PAID_USERS Slack 힌트를 보고 개선
- CONVERSION_RATE 구매유저 / 총유저 계산식으로 안한 것 개선
- 원래 했었는데 최종 SQL을 고치는 과정에서 CONVERSION_RATE를 잘못 썼었습니다.

keeyong commented:
마지막 커멘트는 3번 문제를 아주 간략하게 바꿀 수 있다는 겁니다. 지금 아래와 같은 SQL이 있는데

```
              select
                  C.CHANNELNAME
                , left(ST.TS, 7)                                                   as YEAR_MONTH
                , count(distinct USERID)                                           as UNIQUE_USERS
                , count(distinct case when AMOUNT > 0 then USERID end)             as PAID_USERS
                , sum(case when REFUNDED is true then AMOUNT * -1 else AMOUNT end) as GROSS_REVENUE
                , sum(case when REFUNDED is false then AMOUNT else 0 end)          as NET_REVENUE
              from RAW_DATA.CHANNEL                        C
                   left join RAW_DATA.USER_SESSION_CHANNEL USC
                             on C.CHANNELNAME = USC.CHANNEL
                   left join RAW_DATA.SESSION_TRANSACTION  STR on USC.SESSIONID = STR.SESSIONID
                   left join RAW_DATA.SESSION_TIMESTAMP    ST on USC.SESSIONID = ST.SESSIONID
              group by 1, 2
              order by 1, 2
```

사실 이걸 조금 바꾸면 이것만으로 답이 나옵니다. 아래 보여드릴께요:

```
              select
                  C.CHANNELNAME
                , left(ST.TS, 7)                                                   as YEAR_MONTH
                , count(distinct USERID)                                           as UNIQUE_USERS
                , count(distinct case when AMOUNT > 0 then USERID end)             as PAID_USERS
                , sum(AMOUNT)                                                      as GROSS_REVENUE
                , sum(case when REFUNDED is false then AMOUNT else 0 end)          as NET_REVENUE
                , ROUND(
                    count(distinct case when AMOUNT > 0 then USERID end)::decimal*100/NULLIF(count(distinct USERID), 0)
                    ,2
                  ) as CONVERSION_RATE
              from RAW_DATA.CHANNEL                        C
                   left join RAW_DATA.USER_SESSION_CHANNEL USC
                             on C.CHANNELNAME = USC.CHANNEL
                   left join RAW_DATA.SESSION_TRANSACTION  STR on USC.SESSIONID = STR.SESSIONID
                   left join RAW_DATA.SESSION_TIMESTAMP    ST on USC.SESSIONID = ST.SESSIONID
              group by 1, 2
              order by 1, 2
```

Conversion rate을 계산하는 부분에서 NULLIF 부분 잘 보세요. 이게 0으로 나누는 것을 방지하기 위함인데 이게 가능한 이유는 SQL에서는 NULL로 나눈 경우가 NULL이 되기 때문입니다.

mingyu-lee committed:
2번 문제 피드백 반영, 3번 문제 GROSS_REVENUE 개선

mingyu-lee committed:
2번 문제 GROSS_REVENUE로 변경

mingyu-lee commented:
@keeyong 

3번 개선하여 다시 PR 올렸습니다.
* PAID_USERS 피드백 반영
* CONVERSION_RATE 제대로 작성
* GROSS_REVENUE 개선

2번의 경우 애초에 GROSS_REVENUE를 구하는 것이 문제인데, NET_REVENUE를 구했네요. (REFUNDED 포함을 잘 못 이해했습니다.)
그런데 피드백 주신 쿼리를 돌려보니 NET_REVENUE 결과대로 해주신 것 같아서 결국 결과는 같은 것 같네요. 
(올려주신 쿼리는 정렬이 nulls first가 기본으로 적용되기 때문에 order by 2 desc nulls last를 하거나 sum 구분에 else 0 을 추가해야 하더군요.)

구문을 분석해보니 제가 기존에 썼던 내용은
REFUNDED true일 때 AMOUNT * -1 (환불처리..) 하고 아닐 경우 AMOUNT 합 계산하고
기영님이 말씀해주신 부분은 단순하게 REFUNDED가 false인 것만 AMOUNT 합 계산하는 거네요.

복잡하게 생각해서 접근 했던 것 같네요

keeyong reviewed (approved):
제가 올린 3번 간단히 한 SQL 꼭 보세요! 수고 많이 하셨습니다.

mingyu-lee committed:
최종 피드백 반영
